File: practice/algorithms/implementation/jumping_on_the_cloud_revisited/jumping_on_the_cloud_revisited.hs
module Main where

solve :: [Int] -> Int -> Int -> Int -> Int
solve cloud energy steps k = solve' cloud energy steps k 1
    where
        solve' cloud energy steps k pos
            | pos < steps = solve' cloud (energy - 2 * d) steps k (pos + 1)
            | otherwise   = energy
            where
                d = cloud !! (pos * k)

main :: IO ()
main = do
    params <- getLine
    cloud' <- getLine
    let [n, k] = map (read :: String -> Int) $ words params
        cloud  = map (read :: String -> Int) $ words cloud'
        steps  = n `div` k
        init_e = if cloud !! 0 == 1 then 2 else 0
        energy = 100 - steps - init_e
    print $ solve cloud energy steps k

File: practice/algorithms/implementation/jumping_on_the_cloud_revisited/jumping_on_the_cloud_revisited.cpp
#include <vector>
#include <iostream>

using namespace std;

int main()
{
    int n, k, h, e, start;
    cin >> n >> k;
    vector<int> cloud(n);
    for (int pos = 0; pos < n; ++pos)
    {
       cin >> cloud[pos];
    }
    
    h = n / k;
    e = 100 - h;
    for (int pos = 1; pos <  h; ++pos)
    {
        e -= 2 * cloud[pos * k];
    }
    
    if (cloud[0])
    {
        e -= 2;
    }
    
    cout << e << endl;
    
    return 0;
}

